org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java

```diff
@@ -0,0 +1,747 @@
+                qubit.setConnectedQubits(connectedQubits);
+            }
+        } else {
+            if (device.getNumberQubits() == null) {
+                logger.error("For device {} of provider {} the number of qubits property is null.", device.getDeviceName(), provider);
+                return;
+            }
+            // for simulators and QPUs with one qubit no coupling map exists, therefore just add the qubits
+            for (int i = 0; i < device.getNumberQubits().intValue(); i++) {
+                Qubit qubit = new Qubit();
+                qubit.setQpu(qpu);
+                qubit.setName(String.valueOf(i));
+                qubit = qubitRepository.save(qubit);
+
+                qubits.put(qubit.getName(), qubit);
+                qpu.getQubits().add(qubit);
+            }
+        }
+    }
+
+    private Qubit constructQubit(QPU qpu, String qubitName) {
+        Qubit qubit = new Qubit();
+        qubit.setQpu(qpu);
+        qubit.setName(qubitName);
+        return qubitRepository.save(qubit);
+    }
+
+    public void addGatesFromDevice(List<String> gateNames, QPU qpu, AWSDevice device) {
+        // add gate to each qubit in the coupling if it operates on multiple qubits
+        List<String> oneQubitGates = gateNames.parallelStream().filter(this::is1QubitGateQasm).collect(Collectors.toList());
+        logger.debug("1 Qubit gate list {}", oneQubitGates);
+        List<String> twoQubitGates = gateNames.parallelStream().filter(Predicate.not(this::is1QubitGateQasm)).collect(Collectors.toList());
+        logger.debug("2 Qubit gate list {}", twoQubitGates);
+        HashMap<Integer, Qubit> qubits = new HashMap<>();
+        for (Integer qubitId : device.getConnectivityMap().keySet()) {
+            Optional<Qubit> optQubit = qubitRepository.findByQpuAndName(qpu, String.valueOf(qubitId));
+            if (optQubit.isEmpty()) {
+                logger.error("Qubit {} is null for device {}", qubitId, device.getDeviceName());
+                continue;
+            }
+            qubits.put(qubitId, optQubit.get());
+        }
+        // Collect all distinct qubit ids
+        if (device.getConnectivityMap() != null) {
+            for (Integer sourceQubitId : device.getConnectivityMap().keySet()) {
+                Qubit sourceQubit = qubits.get(sourceQubitId);
+                for (String gateName : oneQubitGates) {
+                    Gate gate = new Gate();
+                    gate.setName(gateName);
+                    gate.setQpu(qpu);
+                    // Add gate to supported gates of the qubit it operates on
+                    sourceQubit.getSupportedGates().add(gate);
+                    gate.addOperatingQubit(sourceQubit);
+                    qubitRepository.save(sourceQubit);
+                    // TODO: This leads to an illegal state representation and it appears to work anyway. Not sure if we need it? (i am not familiar enough with JPA)
+                    // gateRepository.save(gate);
```

Comment:
@wederbn Here I got an issue with the persistence of the gates  using JPA. However in my testing the model gets updated correctly but if I include the line an exception gets thrown. What am I doing wrong?

Reply:
The supportedGates has the annotation to cascade the changes in connected entities, i.e., gates in this case. This means the qubitRepository.save() operation also commits the newly created gate to the DB and assigns it a corresponding ID. gateRepository.save(gate); tries to add the gate to the DB again, thus, the illegal state appears. So just storing the qubit should be enough